### PR TITLE
Band Multiplexing

### PR DESCRIPTION
**Pull request by jbouffard**
This PR adds the `read` method to `RasterSourceRDD`. This method will take in any number of `ReadingSource`s and convert them into a single `MultibandTile` based on the order of the bands given.

### CODE REVIEW

File: vlm/src/main/scala/geotrellis/contrib/vlm/RasterSourceRDD.scala [hunk 1]
```diff
@@ -32,6 +32,84 @@ import scala.reflect.ClassTag
 object RasterSourceRDD {
   final val PARTITION_BYTES: Long = 128l * 1024 * 1024
 
+  def read(
+    readingSources: Seq[ReadingSource],
+    layout: LayoutDefinition,
+    partitionBytes: Long = PARTITION_BYTES
+  )(implicit sc: SparkContext): MultibandTileLayerRDD[SpatialKey] = {
+    val cellTypes = readingSources.map { _.source.cellType }.toSet
+    require(cellTypes.size == 1, s"All RasterSources must have the same CellType, but multiple ones were found: $cellTypes")
```
[the remark this answers is not in the document]
Comment (jbouffard) [reply vlm/src/main/scala/geotrellis/contrib/vlm/RasterSourceRDD.scala]: I agree. I was thinking that as well after I wrote it. It's a requirement we have elsewhere too: https://github.com/geotrellis/geotrellis-contrib/blob/master/vlm/src/main/scala/geotrellis/contrib/vlm/RasterSourceRDD.scala#L41-L48

File: vlm/src/main/scala/geotrellis/contrib/vlm/RasterSourceRDD.scala [hunk 2]
```diff
@@ -32,6 +32,84 @@ import scala.reflect.ClassTag
...
+
+    val result: RDD[(SpatialKey, MultibandTile)] =
+    groupedSourcesRDD.mapValues { iter =>
+      val mappedBands: Map[Int, Option[MultibandTile]] = iter.toSeq.sortBy { _._1 }.toMap
+
+      val tiles: Seq[Tile] =
+        targetIndexes.map { index =>
+          mappedBands.get(index).getOrElse(None) match {
+            case Some(multibandTile) => multibandTile.band(0)
+            case None => noDataTile
```
[the remark this answers is not in the document]
Comment (jbouffard) [reply vlm/src/main/scala/geotrellis/contrib/vlm/RasterSourceRDD.scala]: Oh, I agree. I think we should definitely make it a `def` if it'll help performance. And I'll change the `mapValues` to `mapPartitions` :+1:

**Comment by pomadchin** [comment]
@jbouffard sounds like a good overload for me. What do you think about such an interface?

```scala
case class ReadingSource(
  source: RasterSource,
  sourceBands: Seq[Int],
  targetBand: Seq[Int]
) extends Serializable
```

Or it would be too complicated? 

And with your suggested overload it can be easily used for singlebands.

**Comment by jbouffard** [comment]
@pomadchin I found that having `sourceBands` and `targetBands` being `Seq[Int]` complicated a lot of the logic and made it harder to reason about. We might want to

File: vlm/src/main/scala/geotrellis/contrib/vlm/RasterSourceRDD.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (echeipesh) [reply vlm/src/main/scala/geotrellis/contrib/vlm/RasterSourceRDD.scala]: The problem here is that the tiles are going to end up in `MultibandTile` which supports only a single `CellType`. So if that's not the case we would have to convert all the bands to matching cell type.

**Comment by echeipesh** [comment]
Just to record my point: `RasterSource` is certainly multi-banded. So some kind of mapping from multiple bands to multiple bands is required here.

File: vlm/src/main/scala/geotrellis/contrib/vlm/RasterSourceRDD.scala [hunk 3]
```diff
@@ -32,6 +32,88 @@ import scala.reflect.ClassTag
...
+    val layerKeyBounds = KeyBounds(mapTransform(combinedExtents))
+    val tileSize = layout.tileCols * layout.tileRows * cellType.bytes
+
+    val layerMetadata =
+      TileLayerMetadata[SpatialKey](cellType, layout, combinedExtents, crs, layerKeyBounds)
+
+    def getNoDataTile = ArrayTile.alloc(cellType, layout.tileCols, layout.tileRows).fill(NODATA).interpretAs(cellType)
+
+    val maxIndex = readingSources.map { _.sourceToTargetBand.values.max }.max
+    val targetIndexes: Seq[Int] = 0 to maxIndex
```
[the remark this answers is not in the document]
Comment (jbouffard) [reply vlm/src/main/scala/geotrellis/contrib/vlm/RasterSourceRDD.scala]: Any index that is missing gets replaced with a `NoDataTile` https://github.com/geotrellis/geotrellis-contrib/blob/201dada2cb731750c3c1c51a6dba8f66d3ee4f4f/vlm/src/main/scala/geotrellis/contrib/vlm/RasterSourceRDD.scala#L101-L104

I also have a test for it: https://github.com/geotrellis/geotrellis-contrib/blob/201dada2cb731750c3c1c51a6dba8f66d3ee4f4f/vlm/src/test/scala/geotrellis/contrib/vlm/RasterSourceRDDSpec.scala#L342

**Pull request merged**